### PR TITLE
fix: make total optional in paginator

### PR DESCRIPTION
**Pull request by arjunlalb**
## Description
Make paginator work with unknown totals.

Currently, paginator needs a total count to work. Updating this to understand following codes
- totalCount of `-1` = unknown count
- totalCount of `-2` = last page

If count is -1, 
- Show `many` instead of total count.
- Allow next page.

If count is -2,
- Show `last` instead of total count.
- Do not allow next page.

Its upto the consumer to provide appropriate values inorder to use this behaviour correctly.

### Testing
Manually tested.

First Page when count is unknown
Screenshot 2023-07-17 at 11 42 04 AM

Intermediate page when count is unknown
Screenshot 2023-07-17 at 11 42 26 AM

Last page (when count is -2)
Screenshot 2023-07-17 at 11 43 22 AM

### CODE REVIEW

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts [hunk 1]
```diff
@@ -32,21 +38,32 @@ export class TracesTableDataSourceModel extends TableDataSourceModel {
...
       data: response.results,
-      totalCount: response.total
+      // We want to avoid showing real totals for traces table.
+      // Provide the `last` code when results are lesser than the limit (aka, we know there are no more results)
+      // Provide the `unknown` code otherwise
+      totalCount: this.ignoreTotal
+        ? response.results.length < request.position.limit
+          ? PaginatorTotalCode.Last
+          : PaginatorTotalCode.Unknown
+        : response.total
```
Comment (Exac) [inline projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:66]: I think we should add a test for this.

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts [hunk 2]
```diff
@@ -32,21 +38,32 @@ export class TracesTableDataSourceModel extends TableDataSourceModel {
...
+    request: TableDataRequest<SpecificationBackedTableColumnDef>
+  ): TableDataResponse<TableRow> {
     return {
       data: response.results,
-      totalCount: response.total
+      // We want to avoid showing real totals for traces table.
+      // Provide the `last` code when results are lesser than the limit (aka, we know there are no more results)
+      // Provide the `unknown` code otherwise
+      totalCount: this.ignoreTotal
+        ? response.results.length < request.position.limit
```
Comment (adisreyaj) [inline projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:63]: Will the result length be ever greater than limit?

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts [hunk 3]
```diff
@@ -32,21 +38,32 @@ export class TracesTableDataSourceModel extends TableDataSourceModel {
       requestType: TRACES_GQL_REQUEST,
       traceType: this.traceType,
       properties: request.columns.filter(column => column.visible).map(column => column.specification),
-      limit: request.position.limit * 2, // Prefetch 2 pages
+      limit: request.position.limit, // Prefetch 2 pages
```
Comment (adisreyaj) [inline projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:41]: Comment not valid anymore.

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts (hunk 2 above)
Comment (arjunlalb) [reply projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:63]: nope.

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts (hunk 1 above)
Comment (arjunlalb) [reply projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:66]: added

File: projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts (hunk 3 above)
Comment (arjunlalb) [reply projects/observability/src/shared/dashboard/data/graphql/table/traces/traces-table-data-source.model.ts:41]: updated

**Review by adisreyaj — APPROVED** [review]

**Review by Exac — APPROVED** [review]

**Pull request merged**